xdsl/dialects/arm_neon.py

```diff
@@ -94,9 +99,24 @@ class NeonArrangement(StrEnum):
     S = "S"
     H = "H"
 
-    def map_to_num_els(self):
-        map = {"D": 2, "S": 4, "H": 8}
-        return map[self.name]
+    def get_num_els_from_arranagement(self):
+        return map_arrangement_to_num_els[self.name]
+
+    @staticmethod
+    def get_arrangement_from_vec_type(vec_type: VectorType):
+        key = str(vec_type)
+        if key not in map_vec_to_arr:
+            raise ValueError(f"Invalid vector type for ARM NEON: {key}")
+        return map_vec_to_arr[str(vec_type)]
+
+
+map_arrangement_to_num_els = {"D": 2, "S": 4, "H": 8}
+
+map_vec_to_arr: dict[str, NeonArrangement] = {
+    "vector<8xf16>": NeonArrangement.H,
+    "vector<4xf32>": NeonArrangement.S,
+    "vector<2xf64>": NeonArrangement.D,
+}
```

Comment:
A few changes:

1. The convention in Python is to use UPPER_SNAKE_CASE for constants
2. These constants can be private and live on classes for namespacing reasons
3. rename helper to be a bit more consise
4. use property for the O(1) lookup for self
5. move other dict into class also
6. rename static method
7. use .get for optional lookup
8. use vector type as key

```suggestion
    _SIZE_BY_ARRANGEMENT = {"D": 2, "S": 4, "H": 8}

    @property
    def arrangement(self):
        return self._SIZE_BY_ARRANGEMENT[self.name]

    _ARRANGEMENT_BY_TYPE: dict[str, NeonArrangement] = {
        VectorType(f16, (8,)): NeonArrangement.H,
        VectorType(f32, (4,)): NeonArrangement.S,
        VectorType(f64, (2,)): NeonArrangement.D,
    }

    @staticmethod
    def from_arrangement(vec_type: VectorType):
        arrangement = NeonArrangement._ARRANGEMENT_BY_TYPE.get(vec_type)
        if arrangement is None:
            raise ValueError(f"Invalid vector type for ARM NEON: {vec_type}")
        return arrangement
```

Reply:
I think this is going to cause issues because it now treats _SIZE_BY_ARRANGEMENT as a member of the enum (this was why I previously had it outside the class or within a method). I can't find any other way of storing it that does not lead to errors.

Reply:
In addition, when trying to use NeonArrangement as the dictionary value type at this point in the code, it complains that "NeonArrangement" is not defined, I guess because the class has not been fully initialized?

Reply:
Yeah that makes sense I didn't realise that enums broke this pattern, it should be fine outside of the class, but below the class definition. It makes sense that NeonArrangement would not be defined if the dictionary is placed above the definition in the file, as the file is executed top to bottom

Reply:
this name is not correct: ```def from_arrangement(vec_type: VectorType):``` as the arrangement is the return value so I propose to change it to from_vec_type (will push this with next commit)

Reply:
Oh yeah that was a brain fart from me your name is great (although maybe `from_vector_type`?)